grobid-core/src/main/java/org/grobid/core/engines/ReferenceSegmenterParser.java

```diff
@@ -636,68 +644,80 @@ else if (n == block.getEndToken()) {
 					startblock = false;
 					endblock = false;
 				}
-				// set corresponding token 
+				// set corresponding token
 	            if (block != null)
 					tokens = block.getTokens();
-				else 
+				else
 					tokens = null;
-				
-				String text = tokenizations.get(n);
-
-                if (text == null) {
-                    continue;
-                }
 
-                if (text.equals("\n") || text.equals("\r")) {
+				if (text.equals("\n") || text.equals("\r")) {
                     newline = true;
                     previousNewline = true;
                     nn = 0;
 					currentLineLength = 0;
 					currentLineProfile = null;
 					//lineIndex++;
-					//currentLineLength = lineLengths.get(lineIndex); 
+					//currentLineLength = lineLengths.get(lineIndex);
                     continue;
-                } 
+                }
 				else {
                     newline = false;
 					nn += text.length(); // +1 for segmentation symbol
 				}
-				
+
 				if (text.equals(" ") || text.equals("\t")) {
                     nn++;
                     continue;
 				}
-				
+
 				if (text.trim().length() == 0) {
 					continue;
 				}
-				
-				LayoutToken token = null;
-				if (tokens != null) {
-					while (tokenIndex < tokens.size()) {
-						token = tokens.get(tokenIndex);
-						if (text.equals(token.getText()))
-							break;
-						tokenIndex++;
-					}
-				}
-				
+
+                LayoutToken token = null;
+                if (tokens != null) {
+                    int i = tokenIndex;
+                    while (i < tokens.size()) {
+                        token = tokens.get(i);
+                        if (text.equals(token.getText())) {
+                            tokenIndex = i;
+                            break;
+                        }
+                        i++;
+                    }
+                }
+
                 if (previousNewline) {
                     newline = true;
                     previousNewline = false;
+					if (token != null && previousFeatures != null) {
```

Comment:
The indentation logic.  Compare start of line to previous line, change indentation if the difference is more than a character width.